torch/csrc/autograd/python_variable.cpp

```diff
@@ -134,70 +167,106 @@ PyObject * THPVariable_Wrap(Variable var)
       (PyTypeObject*)THPVariableClass, std::move(var), status);
 }
 
-static int THPVariable_traverse(THPVariable *self, visitproc visit, void *arg)
-{
-  Py_VISIT(self->backward_hooks);
-  // We don't want to traverse the grad_fn, even if the Variable owns it and the
-  // shared pointer's use count is 1. This is because we would need to treat
-  // the grad_fn as part of the Python state and hold the GIL sometimes when
-  // grad_fn's shared_ptr is copied, otherwise a race condition with the Python
-  // GC could occur. Holding the GIL when the shared_ptr is copied adds
-  // undesirable complexity/overhead.
-  //
-  // When hooks, a Variable, and its grad_fn are involved in a Python reference
-  // cycle, because we're not traversing the grad_fn, the reference cycle will
-  // in fact leak.
-  //
-  // See https://gist.github.com/zou3519/7ac92b84dd7d206dcc6eae55fee8372c
-  // for more details about the race condition involving traversing the grad_fn
-  // and the python GC.
+static int THPVariable_clear(THPVariable* self) {
+  Py_CLEAR(self->backward_hooks);
   const auto& tensor = THPVariable_Unpack(self);
   if (tensor.defined()) {
-    for (const auto& hook : torch::autograd::impl::hooks(tensor)) {
-      if (auto pyhook = dynamic_cast<PyFunctionPreHook*>(hook.get())) {
-        Py_VISIT(pyhook->dict);
+    // Two situations to consider:
+    //    PyObject -owns-> Tensor
+    //        unsafeIsBorrowed() is FALSE.  We're obligated to look through
+    //        Tensor to break references.  Clearing cdata could induce the
+    //        destruction of the C++ Tensor.
```

Comment:
Actually it has to induce it.
If there were other references to the c++ Tensor, we would not let the python object die and would resurrect it (and so we would not end up in this function).